src/common/gpio_servo.rs

```diff
@@ -0,0 +1,191 @@
+//! A generalized servo implementation that uses a PWM signal on a GPIO pin
+//! via methods on a board component
+//!
+//! # Creating a GPIO pin servo and turning it 90 degrees
+//!
+//! ```ignore
+//!
+//! let board = FakeBoard::new(vec![]);
+//!
+//! let servo_settings = GpioServoSettings {
+//!     min_angle_deg: 0,
+//!     max_angle_deg: 180,
+//!     min_period_us: 500,
+//!     max_period_us: 2500,
+//!     pwm_resolution: 0
+//! }
+//!
+//! let mut servo = GpioServo::new(board, 12, servo_settings);
+//!
+//! servo.move_to(90).unwrap()
+//!
+//! ```
+
+use std::sync::{Arc, Mutex};
+
+use super::{
+    actuator::Actuator,
+    board::{Board, BoardType},
+    config::{AttributeError, ConfigType},
+    registry::{get_board_from_dependencies, ComponentRegistry, Dependency},
+    servo::{Servo, ServoType},
+    status::Status,
+};
+
+pub(crate) fn register_models(registry: &mut ComponentRegistry) {
+    if registry.register_servo("gpio", &from_config).is_err() {
+        log::error!("gpio model is already registered")
+    }
+}
+
+#[allow(dead_code)]
+pub(crate) fn from_config(
+    cfg: ConfigType,
+    dependencies: Vec<Dependency>,
+) -> anyhow::Result<ServoType> {
+    let board = get_board_from_dependencies(dependencies);
+    if board.is_none() {
+        return Err(anyhow::anyhow!(
+            "actual board is required to be passed to configure a gpio servo"
+        ));
+    }
+    let board_unwrapped = board.unwrap();
+    let servo_settings = GpioServoSettings::from_config(&cfg)?;
+    let pin = cfg.get_attribute::<i32>("pin").map_err(|err| match err {
+        AttributeError::ConversionImpossibleError => {
+            anyhow::anyhow!("could not convert pin attribute format when configuring gpio servo")
+        }
+        _ => anyhow::anyhow!("error configuring gpio servo: {:?}", err),
+    })?;
+    Ok(Arc::new(Mutex::new(GpioServo::<BoardType>::new(
+        board_unwrapped,
+        pin,
+        servo_settings,
+    )?)))
+}
+
+#[derive(Debug)]
+pub struct GpioServoSettings {
+    pub min_angle_deg: f64,
+    pub max_angle_deg: f64,
+    pub min_period_us: u32,
+    pub max_period_us: u32,
+    pub frequency: u32,
+    /// when 0, pwm_resolution is not considered when calculating the PWM duty cycle
+    /// necessary to move the servo to particular angular position
+    pub pwm_resolution: u32,
+}
+
+impl GpioServoSettings {
+    pub fn from_config(cfg: &ConfigType) -> anyhow::Result<Self> {
+        let min_angle_deg = cfg
+            .get_attribute::<f64>("min_angle_deg")
+            .unwrap_or_default();
+        let max_angle_deg = cfg.get_attribute::<f64>("max_angle_deg").unwrap_or(180.0);
+        let min_period_us = cfg.get_attribute::<u32>("min_width_us").unwrap_or(500);
+        let max_period_us = cfg.get_attribute::<u32>("max_period_us").unwrap_or(2500);
+        let frequency = cfg.get_attribute::<u32>("frequency_hz").unwrap_or(300);
```

Comment:
Is there a reason for these defaults? Are they just general settings that would work for most gpio servos?
Are these settings visible in Viam or is there a way for the user to see what the implicit settings are after creation?